contrib/IECoreAlembic/src/IECoreAlembic/MeshAlgo.cpp

```diff
@@ -32,61 +33,107 @@
 //
 //////////////////////////////////////////////////////////////////////////
 
-#include "IECore/MeshPrimitive.h"
-
-#include "IECoreAlembic/FromAlembicSubDConverter.h"
+#include "IECoreAlembic/ObjectAlgo.h"
+#include "IECoreAlembic/MeshAlgo.h"
+#include "IECoreAlembic/GeomBaseAlgo.h"
 
 using namespace IECore;
 using namespace IECoreAlembic;
 using namespace Alembic::AbcGeom;
 
-FromAlembicSubDConverter::ConverterDescription<FromAlembicSubDConverter> FromAlembicSubDConverter::g_description;
+namespace IECoreAlembic
+{
 
-IE_CORE_DEFINERUNTIMETYPED( FromAlembicSubDConverter );
+namespace MeshAlgo
+{
 
-FromAlembicSubDConverter::FromAlembicSubDConverter( Alembic::Abc::IObject iSubD )
-	:	FromAlembicGeomBaseConverter( "Converts AbcGeom::ISubD objects to IECore::MeshPrimitive objects", iSubD )
+IECore::MeshPrimitivePtr convert( const Alembic::AbcGeom::IPolyMesh &mesh, const Alembic::Abc::ISampleSelector &sampleSelector )
 {
+	const IPolyMeshSchema &iPolyMeshSchema = mesh.getSchema();
+
+	IPolyMeshSchema::Sample sample = iPolyMeshSchema.getValue( sampleSelector );
+
+	IntVectorDataPtr verticesPerFace = new IntVectorData();
+	verticesPerFace->writable().insert(
+		verticesPerFace->writable().begin(),
+		sample.getFaceCounts()->get(),
+		sample.getFaceCounts()->get() + sample.getFaceCounts()->size()
+	);
+
+	IntVectorDataPtr vertexIds = new IntVectorData();
+	vertexIds->writable().insert(
+		vertexIds->writable().begin(),
+		sample.getFaceIndices()->get(),
+		sample.getFaceIndices()->get() + sample.getFaceIndices()->size()
+	);
+
+	V3fVectorDataPtr points = new V3fVectorData();
+	points->writable().resize( sample.getPositions()->size() );
+	memcpy( &(points->writable()[0]), sample.getPositions()->get(), sample.getPositions()->size() * sizeof( Imath::V3f ) );
```

Comment:
Oh, interesting. Just did a test here and memcpy/copy/insert had basically identical performance. Consolidating it all into one place does sound good.

One wrinkle that I haven't dealt with yet is that it seems like in practice AbcExport in Maya often writes array samples of size 1 when it should be writing scalar samples. For instance, `AlembicInputTest.testBoolGeomParam` is loading BoolVectorData, but the attribute I added in Maya was a single bool. It seems like AbcImport and the Arnold procedural are then using heuristics to say "oh, i think i'll load this array as a scalar value". I think we're going to need similar heuristics too.

If it's OK with you, I'd like to deal with this consolidation in a separate PR after taking a bit of time to figure out what this heuristic should be, and how other packages are dealing with this stuff.